File: zig-mcp-template/src/tools/registry.zig
const std = @import("std");
const mcp = @import("../mcp.zig");
const dice = @import("dice.zig");

/// Function signature for tool execution callbacks
pub const ToolExecuteFn = *const fn (
    allocator: std.mem.Allocator,
    arguments: ?std.json.Value,
) anyerror![]const u8;

/// Definition of a single tool with metadata and execution callback
pub const ToolDefinition = struct {
    name: []const u8,
    description: []const u8,
    schema_json: []const u8,
    execute_fn: ToolExecuteFn,
};

/// Comptime-known array of all registered tools
pub const TOOLS = [_]ToolDefinition{
    .{
        .name = dice.NAME,
        .description = dice.DESCRIPTION,
        .schema_json = dice.SCHEMA,
        .execute_fn = &dice.execute,
    },
};

/// Find a tool by name, returns null if not found
pub fn findToolByName(name: []const u8) ?*const ToolDefinition {
    for (&TOOLS) |*tool| {
        if (std.mem.eql(u8, tool.name, name)) {
            return tool;
        }
    }
    return null;
}

/// Generate an MCP tools list from the registry
pub fn listTools(allocator: std.mem.Allocator) !struct {
    tools: []mcp.Tool,
    schemas: []std.json.Parsed(std.json.Value),
} {
    var tools = try allocator.alloc(mcp.Tool, TOOLS.len);
    errdefer allocator.free(tools);

    var schemas = try allocator.alloc(std.json.Parsed(std.json.Value), TOOLS.len);
    errdefer {
        for (schemas[0..tools.len]) |*schema| {
            schema.deinit();
        }
        allocator.free(schemas);
    }

    for (TOOLS, 0..) |tool_def, i| {
        // Parse the JSON schema for this tool
        const parsed = try std.json.parseFromSlice(
            std.json.Value,
            allocator,
            tool_def.schema_json,
            .{},
        );
        schemas[i] = parsed;

        tools[i] = .{
            .name = tool_def.name,
            .description = tool_def.description,
            .inputSchema = parsed.value,
        };
    }

    return .{
        .tools = tools,
        .schemas = schemas,
    };
}

/// Clean up resources allocated by listTools
pub fn freeToolsList(allocator: std.mem.Allocator, list: anytype) void {
    for (list.schemas) |*schema| {
        schema.deinit();
    }
    allocator.free(list.schemas);
    allocator.free(list.tools);
}

File: cpp20-mcp-template/src/tools/registry.cpp
#include "registry.hpp"
#include <algorithm>

namespace tools {

ToolRegistry& ToolRegistry::instance() {
    static ToolRegistry instance;
    return instance;
}

void ToolRegistry::registerTool(ToolDefinition tool) {
    tools_.push_back(std::move(tool));
}

const ToolDefinition* ToolRegistry::findToolByName(const std::string& name) const {
    auto it = std::find_if(tools_.begin(), tools_.end(),
        [&name](const ToolDefinition& tool) {
            return tool.name == name;
        });

    return (it != tools_.end()) ? &(*it) : nullptr;
}

const std::vector<ToolDefinition>& ToolRegistry::getTools() const {
    return tools_;
}

std::vector<mcp::Tool> ToolRegistry::listTools() const {
    std::vector<mcp::Tool> result;
    result.reserve(tools_.size());

    for (const auto& tool_def : tools_) {
        rapidjson::Document schema;
        schema.Parse(tool_def.schema_json.c_str());

        result.push_back(mcp::Tool{
            .name = tool_def.name,
            .description = tool_def.description,
            .inputSchema = std::move(schema)
        });
    }

    return result;
}

} // namespace tools
